File: python-engine/factory.py
import logging

import gi
gi.require_version('IBus', '1.0')

import engine
from gi.repository import IBus


class EngineFactory(IBus.Factory):
    def __init__(self, bus):
        self.__bus = bus
        super(EngineFactory, self).__init__(self.__bus)

        self.__id = 0
        self.logger = logging.getLogger('thaime.factory')
        self.logger.info("Engine factory created")

    def do_create_engine(self, engine_name):
        self.logger.info(f"Creating engine: {engine_name}")
        self.__id += 1
        engine_path = f"/org/freedesktop/IBus/Thaime/Engine/{self.__id}"
        self.logger.info(f"Creating engine instance with path: {engine_path}")

        if engine_name == "thaime":
            return engine.Engine(self.__bus, engine_path)
        else:
            self.logger.error(f"Unknown engine name: {engine_name}")
            raise NotImplementedError(f"Unknown engine name: {engine_name}")

File: src/factory.rs
use std::sync::Arc;
use log::{info, warn};
use zbus::{Connection, interface, fdo};
use zvariant::ObjectPath;

use crate::engine::ThaiEngine;

// IBus Factory interface
pub struct IBusEngineFactory {
    connection: Arc<Connection>,
    engine_id_counter: std::sync::atomic::AtomicU32,
}

#[interface(name = "org.freedesktop.IBus.Factory")]
impl IBusEngineFactory {
    #[zbus(name = "CreateEngine")]
    async fn create_engine(&self, engine_name: &str) -> fdo::Result<ObjectPath<'static>> {
        info!("=== Factory: create_engine called with engine_name: '{}' ===", engine_name);
        
        if engine_name == "thaime-rust" {
            let engine_id = self.engine_id_counter.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
            let engine_path = format!("/org/freedesktop/IBus/Engine/ThaimeRust/{}", engine_id);
            
            info!("=== Factory: Creating engine instance with path: {} ===", engine_path);
            
            // Create the engine instance
            let thai_engine_core = Arc::new(ThaiEngine::new());
            let ibus_engine = crate::ibus::IBusThaiEngine::new(thai_engine_core);
            
            // Export the engine at the specific path
            let obj_path = ObjectPath::try_from(engine_path.clone()).unwrap();
            match self.connection.object_server().at(&obj_path, ibus_engine).await {
                Ok(_) => {
                    info!("=== Factory: Successfully exported engine at path: {} ===", engine_path);
                    
                    // Return a static ObjectPath
                    let static_path = ObjectPath::try_from(engine_path).unwrap();
                    Ok(static_path)
                }
                Err(e) => {
                    warn!("=== Factory: Failed to export engine at path {}: {} ===", engine_path, e);
                    Err(fdo::Error::Failed(format!("Failed to export engine: {}", e)))
                }
            }
        } else {
            warn!("=== Factory: Unknown engine name: '{}' ===", engine_name);
            Err(fdo::Error::Failed(format!("Unknown engine name: {}", engine_name)))
        }
    }
}

impl IBusEngineFactory {
    pub fn new(connection: Arc<Connection>) -> Self {
        Self {
            connection,
            engine_id_counter: std::sync::atomic::AtomicU32::new(0),
        }
    }
}